velox/benchmarks/SimpleArithmetic.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+#include "velox/expression/tests/VectorFuzzer.h"
+#include "velox/functions/Registerer.h"
+#include "velox/functions/lib/benchmarks/FunctionBenchmarkBase.h"
+#include "velox/functions/prestosql/Arithmetic.h"
+#include "velox/parse/ExpressionsParser.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::test;
+
+namespace {
+
+class SimpleArithmeticBenchmark
+    : public functions::test::FunctionBenchmarkBase {
+ public:
+  SimpleArithmeticBenchmark() : FunctionBenchmarkBase() {
+    registerFunction<functions::MultiplyFunction, double, double, double>(
+        {"multiply"});
+  }
+
+  void setInput(const TypePtr& inputType, const RowVectorPtr& rowVector) {
+    inputType_ = inputType;
+    rowVector_ = rowVector;
+  }
+
+  // Runs `expression` `times` times.
+  void run(const std::string& expression, size_t times) {
+    folly::BenchmarkSuspender suspender;
+    auto exprSet = compileExpression(expression, inputType_);
+    suspender.dismiss();
+
+    size_t count = 0;
+    for (auto i = 0; i < times; i++) {
+      count += evaluate(exprSet, rowVector_)->size();
+    }
+    folly::doNotOptimizeAway(count);
+  }
+
+ private:
+  TypePtr inputType_;
+  RowVectorPtr rowVector_;
+};
+
+SimpleArithmeticBenchmark benchmark;
+
+BENCHMARK(multiply, n) {
+  benchmark.run("multiply(a, a)", n);
+}
+
+BENCHMARK(multiplyHalfNull, n) {
+  benchmark.run("multiply(a, half_null)", n);
+}
+
+BENCHMARK(multiplyConstant, n) {
+  benchmark.run("multiply(a, constant)", n);
+}
+
```

Comment:
I have ran into cases where when i have so many benchmarks in the same process
the order starts effecting results. (folly benchmarks also)
Just keep that in mind, and if it happen we might consider having 1 per executable.

Reply:
Interesting. I'll check it out. Do you have any ideas why that could happen? Because we're capturing wall time instead of actual cpu time?

Reply:
I am not sure, what was happening but it was a case i was looking at with @kevinwilfong .